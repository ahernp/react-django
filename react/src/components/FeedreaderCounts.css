#feedreader-entry-counts {
    width:20%;
    float:left;
}
.selected {
    text-decoration: underline;
}
