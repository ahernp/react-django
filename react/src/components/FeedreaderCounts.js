import React from 'react';

import './FeedreaderCounts.css';

function countEntries(groups, feeds, entries) {
    let feedCounts = {};
    for (let feed of feeds) {
        feedCounts[feed.id] = {count: 0}
    }

    for (let entry of entries) {
        feedCounts[entry.feed_id].count++;
    }

    let groupCounts = {};
    for (let group of groups) {
        groupCounts[group.id] = {count: 0}
    }

    for (let feed of feeds) {
        if (feed.group_id)
            groupCounts[feed.group_id].count += feedCounts[feed.id].count;
    }

    return { groupCounts, feedCounts }
}

export default class FeedreaderCounts extends React.Component {
    render() {
        const { setFilterFeedId, filterFeedId, showUnreadEntries, groups, feeds, entries } = this.props;
        if (feeds === undefined)
			return <p>No data</p>;
        const unreadEntries = entries.filter(entry => !entry.read_flag);
        const { groupCounts, feedCounts } = countEntries(
            groups,
            feeds,
            showUnreadEntries ? unreadEntries : entries);

        const groupList = groups.map(group => {
            const entryList = feeds.filter(feed => feed.group_id === group.id && feedCounts[feed.id].count > 0)
                .map(feed => (
                    <p key={feed.id} className={filterFeedId === feed.id ? 'selected' : ''} onClick={setFilterFeedId(feed.id)}>
                        {feed.title} ({feedCounts[feed.id].count})
                    </p>
                ));
            return (
                <React.Fragment key={group.id}>
                    <h3>{group.name} {groupCounts[group.id].count > 0 && '('+groupCounts[group.id].count+')'}</h3>
                    <div>{entryList}</div>
                </React.Fragment>
            )
        });

        const totalEntries = showUnreadEntries ? unreadEntries.length : entries.length;

        return (
            <div id="feedreader-entry-counts">
                <h2 onClick={setFilterFeedId(undefined)}>All {totalEntries > 0 && '(' + totalEntries + ')'}</h2>
                <div>{groupList}</div>
            </div>
        );
    }
}
