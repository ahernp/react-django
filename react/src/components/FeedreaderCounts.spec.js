import FeedreaderCounts from "./FeedreaderCounts";

const GROUPS = [{id: 0, name: 'Group One'}, {id: 1, name: 'Group Two'}];
const FEEDS = [{id: 0, title: 'Feed One', group_id: 0}, {id: 1, title: 'Feed Two', group_id: 1}];
const ENTRIES = [
    {id: 0, title: 'Entry One', feed_id: 0, read_flag: false},
    {id: 1, title: 'Entry Two', feed_id: 0, read_flag: false},
    {id: 2, title: 'Entry Three', feed_id: 1, read_flag: true},
    {id: 3, title: 'Entry Four', feed_id: 1, read_flag: true}
];

it('renders correctly, count all', () => {
    const wrapper = render(
        <FeedreaderCounts
            setFilterFeedId={() => {}}
            filterFeedId={undefined}
            showUnreadEntries={false}
            groups={GROUPS}
            feeds={FEEDS}
            entries={ENTRIES}
        />
    );

  expect(wrapper).toMatchSnapshot();
});

it('renders correctly, count unread', () => {
    const wrapper = render(
        <FeedreaderCounts
            setFilterFeedId={() => {}}
            filterFeedId={undefined}
            showUnreadEntries={true}
            groups={GROUPS}
            feeds={FEEDS}
            entries={ENTRIES}
        />
    );

    expect(wrapper).toMatchSnapshot();
});
