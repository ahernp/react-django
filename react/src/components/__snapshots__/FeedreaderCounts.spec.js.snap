// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly, count all 1`] = `
<div
  id="feedreader-entry-counts"
>
  <h2>
    All (4)
  </h2>
  <div>
    <h3>
      Group One (2)
    </h3>
    <div>
      <p
        class=""
      >
        Feed One (2)
      </p>
    </div>
    <h3>
      Group Two (2)
    </h3>
    <div>
      <p
        class=""
      >
        Feed Two (2)
      </p>
    </div>
  </div>
</div>
`;

exports[`renders correctly, count unread 1`] = `
<div
  id="feedreader-entry-counts"
>
  <h2>
    All (2)
  </h2>
  <div>
    <h3>
      Group One (2)
    </h3>
    <div>
      <p
        class=""
      >
        Feed One (2)
      </p>
    </div>
    <h3>
      Group Two 
    </h3>
    <div>
      
    </div>
  </div>
</div>
`;
